File: src/futurnal/cli/schema.py
"""Schema Stats CLI commands for schema evolution visualization.

Step 08: Frontend Intelligence Integration - Phase 5

Provides commands for:
- Getting schema statistics (entity types, relationship types, counts)
- Getting schema evolution timeline
- Getting quality metrics

Research Foundation:
- GFM-RAG: Schema-aware graph construction
- ACE: Adaptive schema evolution
"""

import json
import logging
from typing import Optional

from typer import Typer, Option

logger = logging.getLogger(__name__)

schema_app = Typer(help="Schema evolution and statistics commands")


def _get_pkg_database():
    """Factory function to get PKG database instance.

    Lazy initialization to avoid circular imports.
    Returns None if PKG module is not available or Neo4j not running.
    """
    try:
        from futurnal.configuration.settings import bootstrap_settings
        from futurnal.pkg.database import PKGDatabaseManager

        settings = bootstrap_settings()
        manager = PKGDatabaseManager(settings.workspace.storage)
        manager.connect()
        return manager
    except ImportError as e:
        logger.warning(f"PKG database module not available: {e}")
        return None
    except Exception as e:
        logger.warning(f"PKG database connection failed: {e}")
        return None


@schema_app.command("stats")
def get_schema_stats(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get schema statistics including entity types, relationships, and metrics.

    Examples:
        futurnal schema stats
        futurnal schema stats --json
    """
    try:
        db = _get_pkg_database()

        # If PKG database is not available, return placeholder data
        if db is None:
            placeholder_response = {
                "success": True,
                "entityTypes": [
                    {"type": "Document", "count": 0, "firstSeen": None, "lastSeen": None},
                    {"type": "Person", "count": 0, "firstSeen": None, "lastSeen": None},
                    {"type": "Concept", "count": 0, "firstSeen": None, "lastSeen": None},
                ],
                "relationshipTypes": [
                    {"type": "MENTIONS", "count": 0, "confidenceAvg": 0.0},
                    {"type": "RELATES_TO", "count": 0, "confidenceAvg": 0.0},
                ],
                "qualityMetrics": {
                    "precision": 0.0,
                    "recall": 0.0,
                    "temporalAccuracy": 0.0,
                },
                "evolutionTimeline": [],
                "message": "PKG database not initialized. Ingest data to populate schema.",
            }
            if output_json:
                print(json.dumps(placeholder_response))
            else:
                print("Schema Statistics:")
                print("-" * 40)
                print("\nPKG database not initialized.")
                print("Run 'futurnal sources obsidian vault scan' to populate the knowledge graph.")
            return

        # Get entity type statistics
        entity_types = []
        try:
            with db.session() as session:
                result = session.run(
                    """
                    MATCH (n)
                    WITH labels(n) AS types, n
                    UNWIND types AS type
                    WITH type, count(n) AS count,
                         min(n.created_at) AS first_seen,
                         max(n.updated_at) AS last_seen
                    RETURN type, count, first_seen, last_seen
                    ORDER BY count DESC
                    """
                )
                for record in result:
                    entity_types.append({
                        "type": record["type"],
                        "count": record["count"],
                        "firstSeen": record["first_seen"].isoformat() if record["first_seen"] else None,
                        "lastSeen": record["last_seen"].isoformat() if record["last_seen"] else None,
                    })
        except Exception as e:
            logger.warning(f"Failed to get entity types: {e}")

        # Get relationship type statistics
        relationship_types = []
        try:
            with db.session() as session:
                result = session.run(
                    """
                    MATCH ()-[r]->()
                    WITH type(r) AS type, count(r) AS count,
                         avg(r.confidence) AS avg_confidence
                    RETURN type, count, avg_confidence
                    ORDER BY count DESC
                    """
                )
                for record in result:
                    relationship_types.append({
                        "type": record["type"],
                        "count": record["count"],
                        "confidenceAvg": record["avg_confidence"] or 0.0,
                    })
        except Exception as e:
            logger.warning(f"Failed to get relationship types: {e}")

        # Get quality metrics (placeholder - would come from ExperientialLearningPipeline)
        quality_metrics = {
            "precision": 0.85,  # Placeholder
            "recall": 0.78,  # Placeholder
            "temporalAccuracy": 0.92,  # Placeholder
        }

        # Get evolution timeline (recent schema changes)
        evolution_timeline = []
        try:
            with db.session() as session:
                result = session.run(
                    """
                    MATCH (n)
                    WHERE n.created_at IS NOT NULL
                    WITH date(n.created_at) AS day, labels(n) AS types, count(n) AS count
                    RETURN day, types, count
                    ORDER BY day DESC
                    LIMIT 30
                    """
                )
                for record in result:
                    types = record["types"]
                    evolution_timeline.append({
                        "timestamp": record["day"].isoformat() if record["day"] else None,
                        "changeType": "entity_added",
                        "details": f"Added {record['count']} {types[0] if types else 'unknown'} entities",
                    })
        except Exception as e:
            logger.warning(f"Failed to get evolution timeline: {e}")

        if output_json:
            output = {
                "success": True,
                "entityTypes": entity_types,
                "relationshipTypes": relationship_types,
                "qualityMetrics": quality_metrics,
                "evolutionTimeline": evolution_timeline,
            }
            print(json.dumps(output))
        else:
            print("Schema Statistics:")
            print("-" * 40)
            print(f"\nEntity Types ({len(entity_types)}):")
            for et in entity_types[:10]:
                print(f"  {et['type']}: {et['count']} entities")

            print(f"\nRelationship Types ({len(relationship_types)}):")
            for rt in relationship_types[:10]:
                print(f"  {rt['type']}: {rt['count']} (avg conf: {rt['confidenceAvg']:.2f})")

            print("\nQuality Metrics:")
            print(f"  Precision: {quality_metrics['precision']:.2f}")
            print(f"  Recall: {quality_metrics['recall']:.2f}")
            print(f"  Temporal Accuracy: {quality_metrics['temporalAccuracy']:.2f}")

    except Exception as e:
        logger.exception("Get schema stats failed")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")
        raise SystemExit(1)

File: desktop/src-tauri/src/commands/schema.rs
//! Schema Stats IPC commands.
//!
//! Step 08: Frontend Intelligence Integration - Phase 5
//!
//! Research Foundation:
//! - GFM-RAG: Schema-aware graph construction
//! - ACE: Adaptive schema evolution
//!
//! Provides schema evolution statistics from PKG.

use serde::{Deserialize, Serialize};
use tauri::command;

/// Entity type statistics.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityTypeStat {
    #[serde(rename = "type")]
    pub entity_type: String,
    pub count: i32,
    pub first_seen: Option<String>,
    pub last_seen: Option<String>,
}

/// Relationship type statistics.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationshipTypeStat {
    #[serde(rename = "type")]
    pub rel_type: String,
    pub count: i32,
    pub confidence_avg: f64,
}

/// Quality metrics.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QualityMetrics {
    pub precision: f64,
    pub recall: f64,
    pub temporal_accuracy: f64,
}

/// Schema evolution event.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaEvolutionEvent {
    pub timestamp: Option<String>,
    pub change_type: String,
    pub details: String,
}

/// Response from schema stats command.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaStatsResponse {
    pub success: bool,
    pub entity_types: Vec<EntityTypeStat>,
    pub relationship_types: Vec<RelationshipTypeStat>,
    pub quality_metrics: QualityMetrics,
    pub evolution_timeline: Vec<SchemaEvolutionEvent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get schema statistics.
///
/// Step 08: Schema Evolution Dashboard
///
/// Calls: `futurnal schema stats --json`
#[command]
pub async fn get_schema_stats() -> Result<SchemaStatsResponse, String> {
    let args = vec!["schema", "stats", "--json"];

    let response: SchemaStatsResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Schema stats CLI failed: {}", e);
            return Ok(SchemaStatsResponse {
                success: false,
                entity_types: vec![],
                relationship_types: vec![],
                quality_metrics: QualityMetrics {
                    precision: 0.0,
                    recall: 0.0,
                    temporal_accuracy: 0.0,
                },
                evolution_timeline: vec![],
                error: Some(format!("Schema stats failed: {}", e)),
            });
        }
    };

    log::info!(
        "Retrieved schema stats: {} entity types, {} relationship types",
        response.entity_types.len(),
        response.relationship_types.len()
    );

    Ok(response)
}
